### DISCUSSION TITLE
Using NamedTuples as Object Storage

### DISCUSSION THREAD

**Discussion by julian-r** [jsvine/pdfplumber#816] [Ideas]
What are your thoughts on replacing all `T_obj` with Named Tuples with a legacy getter layer? 
I think that would give a great benefit in 
- type safety. 
- The immutability of tuples here can also be an advantage. 
- Will probably affect memory efficiency
- Might improve performance a little

I would start implementing these things if you consider merging such changes.

**Comment by jsvine** [comment]
Thank you for this intriguing proposal, @julian-r! I've thought about something similar over the years, but have worried about backwards compatibility. Your proposed legacy getter layer certainly helps with that. Sub-concerns:
 
1. How large would the performance reduction be for people using the legacy getters?

2. Could the legacy getter layer handle use-cases such as `pandas.DataFrame(page.chars)`, and other similar calls where the code expects a `dict` but isn't explicitly calling, e.g., `obj["x0"]`? I get the sense that this is possible on a library-by-library basis, but is there a more general solution? I haven't looked deeply into that.

I'm certainly open to exploring your proposal. Is there a small-scale experiment that we could do as a proof-of-concept, before trying to implement the changes across the entire library?

**Reply by jsvine** [reply]
Minor update re. 2, based on a quick experiment: It seems that `pandas.DataFrame` handles lists of namedtuples well, which is reassuring.
